semanticdb/scalac/library/src/main/scala/scala/meta/internal/semanticdb/scalac/TextDocumentOps.scala

```diff
@@ -385,62 +381,165 @@ trait TextDocumentOps { self: SemanticdbOps =>
+                        )
+                      )
+                    case gfun if isForSynthetic(gfun) =>
+                      val range = gimpl.pos.toMeta.toRange
+                      val synthTree = forSyntheticOrOrig(gimpl)
+                      synthetics += s.Synthetic(
+                        range = Some(range),
+                        tree = synthTree
+                      )
+                    case gfun =>
+                      synthetics += s.Synthetic(
+                        range = Some(gfun.pos.toMeta.toRange),
+                        tree = s.ApplyTree(
+                          fn = s.OriginalTree(
+                            range = Some(gfun.pos.toMeta.toRange)
+                          ),
+                          args = gimpl.args.map(_.toSemanticTree)
+                        )
+                      )
+                  }
+                case g.TypeApply(fun, targs @ List(targ, _*)) =>
+                  if (targ.pos.isRange) return
+                  // for loops
+                  val fnTree = fun match {
+                    case ApplySelect(select @ g.Select(qual, nme)) if isSyntheticName(select) =>
+                      visitedSyntheticParent += select
+                      visitedSyntheticParent += fun
+                      val symbol = select.symbol.toSemantic
+                      s.SelectTree(
+                        qual = s.OriginalTree(range = Some(qual.pos.toMeta.toRange)),
+                        id = Some(s.IdTree(sym = symbol))
+                      )
+                    case _ =>
+                      s.OriginalTree(
+                        range = Some(fun.pos.toMeta.toRange)
+                      )
+                  }
+                  synthetics += s.Synthetic(
+                    range = Some(fun.pos.toMeta.toRange),
+                    tree = s.TypeApplyTree(
+                      fn = fnTree,
+                      targs = targs.map(_.tpe.toSemanticTpe)
+                    )
+                  )
+                case ApplySelect(select @ g.Select(qual, nme)) if isSyntheticName(select) =>
+                  visitedSyntheticParent += select
+                  val symbol = select.symbol.toSemantic
+                  synthetics += s.Synthetic(
+                    range = Some(qual.pos.toMeta.toRange),
+                    tree = s.SelectTree(
+                      qual = s.OriginalTree(range = Some(qual.pos.toMeta.toRange)),
+                      id = Some(s.IdTree(sym = symbol))
+                    )
+                  )
+                case gtree if isForSynthetic(gtree) =>
+                  val range = gtree.pos.toMeta.toRange
+                  val synthTree = forSyntheticOrOrig(gtree)
+                  synthetics += s.Synthetic(
+                    range = Some(range),
+                    tree = synthTree
+                  )
+                case _ =>
+                // do nothing
+              }
```

Comment:
This looks plausible as far as my mental debugging goes. I can't guarantee that this is 100% correct, since the for loop desugaring in Scalac is way too involved, but we have tests for that.

Reply:
It is not correct with `=` statements, but I decided that this was out of scope as it was not tested in current tests. For example:

```scala
for {
  x <- 1 to 10
  y = x
} yield x + y
```

The following is produced:
```scala
for (x <- 1 to 10; y = x) yield x + y =>
orig(1 to 10)
.map[Tuple2[Int, Int], IndexedSeq[Tuple2[Int, Int]]]({ (x) =>
    orig(y = x)
})(canBuildFrom[Tuple2[Int, Int]])
.map[Int, IndexedSeq[Int]]({ (local10) =>
    orig(x + y)
})(canBuildFrom[Int])
```

Which is not terrible, but is different from expected:

```scala
orig(1 to 10)
.map[(Int, Int), scala.collection.immutable.IndexedSeq[(Int, Int)]](
    ((x: Int) => {
    val y = x;
    scala.Tuple2.apply[Int, Int](x, y)
    }))(scala.collection.immutable.IndexedSeq.canBuildFrom[(Int, Int)])
.map[Int, Any](((x$1: (Int, Int)) =>
    (x$1: @scala.unchecked) match {
    case scala.Tuple2((x @ _), (y @ _)) => orig(x + y)
    }))(scala.collection.immutable.IndexedSeq.canBuildFrom[Int])
```
Although to fully support this, we'd need to add a match ADT node or something to the synthetics tree..

Reply:
Please submit a ticket.

Reply:
Created #1711